### DISCUSSION TITLE
ScriptReference/MonoBehaviour.FixedUpdate

### DISCUSSION THREAD

**Discussion by giscus** [Developer-Notes-Extension/Unity-Notes#144] [Notes]
# ScriptReference/MonoBehaviour.FixedUpdate

https://docs.unity3d.com/ScriptReference/MonoBehaviour.FixedUpdate.html

**Comment by vertxxyz** [comment]
Do not poll instantaneous input functions in `FixedUpdate`. It doesn't run in sync with `Update` or `LateUpdate`, so polling functions like [`GetKeyDown`](https://docs.unity3d.com/ScriptReference/Input.GetKeyDown.html) will inconsistently register.  
Instead store the state as polled from `Update` and [consume it](https://unity.huh.how/programming/input/built-in-input/input-in-fixed-update) here as necessary.

**Comment by vertxxyz** [comment]
See [Important Classes - Time](https://docs.unity3d.com/Manual/TimeFrameManagement.html)'s **Fixed Timestep** for another detailed writeup with diagrams and graphs on how fixed update works.

**Comment by akela-morse** [comment]
`FixedUpdate` runs prior to the internal physics update so that you can have a chance to change the properties of `Rigidbody` and other Physics components before they get written to `Transform`.

But there are specific cases where you may want to run some code **after** the physics update. You can fake a “LateFixedUpdate” by writing `FixedUpdate` as an asynchronous [`Awaitable`](https://docs.unity3d.com/ScriptReference/Awaitable.html) and await for [`FixedUpdateAsync`](https://docs.unity3d.com/6000.0/Documentation/ScriptReference/Awaitable.FixedUpdateAsync.html). This will effectively run the code **after** the internal physics update and after animations are evaluated (provided the `Animator` has been set to update on the fixed loop).

Be aware that this trick, while useful in some very specific cases (such as handling moving platforms on the fixed loop), can potentially cause some serious issues if you try to query the physics engine at this time. Double-check the full list of internal functions on the [execution order](https://docs.unity3d.com/6000.0/Documentation/Manual/execution-order.html) to make sure you're not doing anything wrong. Always update your `Rigidbody` components in the normal `FixedUpdate` loop.
